pkg/chunk/elastic/es_index_client.go

```diff
@@ -0,0 +1,316 @@
+
+	bulkRequest := e.client.Bulk()
+	for _, entry := range b.entries {
+		index := IndexEntry{Hash: entry.HashValue, Range: string(entry.RangeValue), Value: string(entry.Value)}
+		req := elastic.NewBulkIndexRequest().Index(indexName).Type(e.cfg.IndexType).Doc(index)
+		bulkRequest = bulkRequest.Add(req)
+	}
+
+	err := instrument.CollectedRequest(ctx, "ES.BatchWrite", esRequestDuration,
+		instrument.ErrorCode, func(ctx context.Context) error {
+			var err error
+			_, err = bulkRequest.Do(ctx)
+			return err
+		})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// readBatch represents a batch of rows read from ElasticSearch.
+type readBatch struct {
+	rangeValue []byte
+	value      []byte
+}
+
+func (r readBatch) Iterator() chunk.ReadBatchIterator {
+	return &elasticReadBatchIterator{
+		readBatch: r,
+	}
+}
+
+type elasticReadBatchIterator struct {
+	consumed bool
+	readBatch
+}
+
+func (r *elasticReadBatchIterator) Next() bool {
+	if r.consumed {
+		return false
+	}
+	r.consumed = true
+	return true
+}
+
+func (r *elasticReadBatchIterator) RangeValue() []byte {
+	return r.rangeValue
+}
+
+func (r *elasticReadBatchIterator) Value() []byte {
+	return r.value
+}
+
+func (e *esClient) QueryPages(ctx context.Context, queries []chunk.IndexQuery, callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)) error {
+	return chunk_util.DoParallelQueries(ctx, e.query, queries, callback)
+}
+
+func (e *esClient) query(ctx context.Context, query chunk.IndexQuery, callback func(chunk.ReadBatch) (shouldContinue bool)) error {
+	var rangeQuery *elastic.RangeQuery
+	var valueTermQuery *elastic.TermQuery
+
+	level.Debug(util.Logger).Log("msg", fmt.Sprintf(
+		"hash [%s], rangeValuePrefix [%s], rangeValueStart [%s]", query.HashValue, query.RangeValuePrefix, query.RangeValueStart))
+	hashTermQuery := elastic.NewTermQuery("hash", query.HashValue)
+	switch {
+	case len(query.RangeValuePrefix) > 0 && query.ValueEqual == nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValuePrefix)).
+			Lt(string(query.RangeValuePrefix) + null)
+
+	case len(query.RangeValuePrefix) > 0 && query.ValueEqual != nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValuePrefix)).
+			Lt(string(query.RangeValuePrefix) + null)
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case len(query.RangeValueStart) > 0 && query.ValueEqual == nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValueStart))
+
+	case len(query.RangeValueStart) > 0 && query.ValueEqual != nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValueStart))
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case query.ValueEqual != nil:
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case query.ValueEqual == nil:
+		break
+	}
+
+	// Build the query
+	baseQuery := e.client.Scroll(query.TableName).Size(e.cfg.MaxFetchDocs).
```

Comment:
Scroll is not the correct API to use for these queries. It certainly can return a large number of results, but it isn't meant for real-time user queries. 

https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html#request-body-search-scroll

See https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html#request-body-search-from-size instead for how to do this with the `search` endpoint. Note there is also limitations here, based on the configuration of ES itself.

Reply:
Considering big amount of data in my case of testing, I prefer to use scroll rather than search-from-size.

Reply:
let me try search_after. https://www.elastic.co/guide/en/elasticsearch/reference/master/search-request-body.html#request-body-search-search-after

Reply:
That sounds like a better approach than Scroll as it saves the stateful nightmare that scroll puts on a cluster under load. However since it is stateless, depending on your caching configuration the query will be executed multiple times. Ordering can change also as new records are added, but I would expect the sort to be on time and the ingesters are writing data chronologically so this shouldn't be as much of an issue.

Reply:
Yeah, considering current approaches: search_from, scroll, search_after, seems `search_after` is the most suitable one for realtime, large data set.